### PR TITLE
Calculate cardinality estimate variance and Simulate number of active registers

### PR DESCRIPTION
**Pull request by jiayu-google**
… registers.

### CODE REVIEW

File: src/simulator/halo_simulator.py [hunk 1]
```diff
@@ -209,6 +210,37 @@ def simulated_reach_by_spend(
         kplus_reaches = ReachPoint.frequencies_to_kplus_reaches(frequencies)
         return ReachPoint(impressions, kplus_reaches, spends)
 
+    def _cardinality_estimate_variance(self, n: int) -> float:
```
Comment (matthewclegg) [inline src/simulator/halo_simulator.py:213]: Perhaps change the name to _liquid_legions_cardinality_estimate_variance?

File: src/simulator/halo_simulator.py [hunk 2]
```diff
@@ -209,6 +210,37 @@ def simulated_reach_by_spend(
...
+        a = self._params.liquid_legions.decay_rate
+        c = a * n / m / (1 - np.exp(-a))
+        variance = (
+            expi(-c) - expi(-c * np.exp(-a)) - expi(-2 * c) + expi(-2 * c * np.exp(-a))
+        )
+        variance *= a * n ** 2 / m / (np.exp(-c) - np.exp(-c * np.exp(-a))) ** 2
+        variance -= n
+        return variance
+
+    def _num_active_registers(
```
Comment (matthewclegg) [inline src/simulator/halo_simulator.py:234]: Add documentation.  Also, perhaps this should be called _liquid_legions_num_active_regions.

**Review by matthewclegg — CHANGES_REQUESTED** [review]

**Review by riemanli — APPROVED** [review]
LGTM.

**Comment by uakyol** [comment]
We are still investigating, how this PR got merged by mistake

**Pull request merged**